nexus/db-queries/src/db/datastore/region.rs

```diff
@@ -118,44 +119,67 @@ impl DataStore {
         volume_id: Uuid,
         disk_source: &params::DiskSource,
         size: external::ByteCount,
+        allocation_strategy: &RegionAllocationStrategy,
     ) -> Result<Vec<(Dataset, Region)>, Error> {
-        // ALLOCATION POLICY
-        //
-        // NOTE: This policy can - and should! - be changed.
-        //
-        // See https://rfd.shared.oxide.computer/rfd/0205 for a more
-        // complete discussion.
-        //
-        // It is currently acting as a placeholder, showing a feasible
-        // interaction between datasets and regions.
-        //
-        // This policy allocates regions to distinct Crucible datasets,
-        // favoring datasets with the smallest existing (summed) region
-        // sizes. Basically, "pick the datasets with the smallest load first".
-        //
-        // Longer-term, we should consider:
-        // - Storage size + remaining free space
-        // - Sled placement of datasets
-        // - What sort of loads we'd like to create (even split across all disks
-        // may not be preferable, especially if maintenance is expected)
-
         let block_size =
             self.get_block_size_from_disk_source(opctx, &disk_source).await?;
         let (blocks_per_extent, extent_count) =
             Self::get_crucible_allocation(&block_size, size);
 
-        let dataset_and_regions: Vec<(Dataset, Region)> =
-            crate::db::queries::region_allocation::RegionAllocate::new(
-                volume_id,
-                block_size.into(),
-                blocks_per_extent,
-                extent_count,
-            )
-            .get_results_async(self.pool())
-            .await
-            .map_err(|e| crate::db::queries::region_allocation::from_pool(e))?;
-
-        Ok(dataset_and_regions)
+        // Right now the SQL queries to allocate regions will shuffle all the
+        // datasets and pick 3. However, if a zpool has multiple datasets, it
+        // may pick 2 datasets on the same zpool. We do not want this, because
+        // then you have 2 datasets on the same disk, and now you don't have
+        // redundancy. We can't deal with that in the SQL query, but we can
+        // *detect* it. So we detect it.
+        //
+        // When a query fails because it picked 2 zpools from the same dataset,
+        // but there were 3 valid zpools it could have picked from,
+        // can_be_retried(error) will return true. It is highly likely that if
+        // we simply retry the query with a different shuffle-ordering, we will
+        // get a valid result. However, we should no retry forever, in the event
+        // that something else is going on, so we limit it.
```

Comment:
> This is sufficient to select dataset which are unique to each sled -- it basically makes candidate datasets "give me the best dataset each sled could offer, by size".

yes, but it also eliminates the random ordering, defeating the point of this PR.

Reply:
Following-up from a discussion in meet:

We should be able to do the following:

1. Use `DISTINCT ON sled` with `ORDER BY sled, random()` to pick random (but valid!) allocation targets from each sled, and
2. Use the output of the subquery from (1) to randomly select sleds to use for provisioning.

Basically: "Randomly pick a valid target from all sleds, then randomly pick one of those sleds"

However, we can do that as a follow-up to this PR.

Reply:
to add on, I'm convinced after talking with sean that it may be possible to do that in the SQL, and we'll find out for sure once we try it. I'll be opening an issue to track picking across multiple sleds. It's important to reach the level of redundancy we want. But, it's potentially more complicated than this change since we may need to wire up a way to allocate within just one sled as well to avoid breaking dev workflows, and we may find more tests that need to be updated. This PR is pretty important to get in sooner than later to avoid the error conditions it addresses, which is why we're keeping allocations across sleds as future work, but we should do that sooner than later.